source/agora/consensus/SlashPolicy.d

```diff
@@ -0,0 +1,286 @@
+
+    Manage the policy for slashing
+
+*******************************************************************************/
+
+public class SlashPolicy
+{
+    // The amount of a penalty
+    public immutable Amount penalty_amount;
+
+    // The address to get a penalty
+    public immutable PublicKey penalty_address;
+
+    // Enrollment manager
+    private EnrollmentManager enroll_man;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            enroll_man = the EnrollmentManager
+            params = the consensus-critical constants
+
+    ***************************************************************************/
+
+    public this (EnrollmentManager enroll_man, immutable(ConsensusParams) params)
+    {
+        this.enroll_man = enroll_man;
+        // TODO: This should be set through ConsensusParams
+        this.penalty_amount = Amount(10_000L * 10_000_000L);
+        this.penalty_address = params.CommonsBudgetAddress;
+    }
+
+    /***************************************************************************
+
+        Get the validators that do not reveal their pre-images and will be
+        slashed through the consensus procoess.
+
+        Params:
+            missing_validators = will contain the validators
+            height = the desired block height to look up the validators for
+
+    ***************************************************************************/
+
+    public void getMissingValidators (ref uint[] missing_validators,
+        in Height height) @safe nothrow
+    {
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        foreach (idx, utxo_key; keys)
+        {
+            if (!checkValidValidator(height, utxo_key))
+            {
+                missing_validators ~= cast(uint)idx;
+            }
+        }
+    }
+
+    /***************************************************************************
+
+        Get the hash of the merkle root of the pre-images for the height
+
+        Params:
+            height = the desired block height to look up the hash for
+
+        Returns:
+            the hash of the merkle root of the pre-images
+
+    ***************************************************************************/
+
+    public Hash getPreimageRoot (in Height height) @safe nothrow
+    {
+        Hash[] keys;
+        if (!this.enroll_man.getEnrolledUTXOs(keys) || keys.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        Hash[] valid_keys;
+        foreach (key; keys)
+        {
+            if (checkValidValidator(height, key))
+                valid_keys ~= key;
+        }
+
+        return this.enroll_man.getRandomSeed(valid_keys, height);
```

Comment:
That's not a merkle root though.

If you want we can just rename it to something like `random_seed` in the `ConsensusData` and in the `Block`. I think it's fine to just store the value of `getRandomSeed()`, it's just that the naming is wrong because that function doesn't actually create a merkle root.